nexus/db-queries/src/db/datastore/volume.rs

```diff
@@ -907,6 +928,66 @@ impl DataStore {
                 }
             })
     }
+
+    /// Record Upstairs repair progress
+    pub async fn upstairs_repair_progress(
+        &self,
+        opctx: &OpContext,
+        upstairs_id: TypedUuid<UpstairsKind>,
+        repair_id: TypedUuid<UpstairsRepairKind>,
+        repair_progress: RepairProgress,
+    ) -> Result<(), Error> {
+        use db::schema::upstairs_repair_notification::dsl as notification_dsl;
+        use db::schema::upstairs_repair_progress::dsl;
+
+        let conn = self.pool_connection_authorized(opctx).await?;
+        let err = OptionalError::new();
+
+        self.transaction_retry_wrapper("upstairs_repair_progress")
+            .transaction(&conn, |conn| {
+                let repair_progress = repair_progress.clone();
+                let err = err.clone();
+
+                async move {
+                    // Check that there is a repair id for the upstairs id
+                    let matching_repair: Option<UpstairsRepairNotification> =
+                        notification_dsl::upstairs_repair_notification
+                            .filter(notification_dsl::repair_id.eq(nexus_db_model::to_db_typed_uuid(repair_id)))
+                            .filter(notification_dsl::upstairs_id.eq(nexus_db_model::to_db_typed_uuid(upstairs_id)))
+                            .filter(notification_dsl::notification_type.eq(UpstairsRepairNotificationType::Started))
+                            .get_result_async(&conn)
+                            .await
+                            .optional()?;
+
+                    if matching_repair.is_none() {
+                        // XXX should be 404
```

Comment:
If this should be 404, can we make it?  Or that's just not possible from this place?

Reply:
Done in 0b37b63bb752243298e894bb186eb8d6e907374b